### ISSUE TITLE
change default value of OOM of Netdata

### ISSUE THREAD

**Issue by luvpreetsingh** [netdata/netdata#2549]
Hi, I am using latest netdata on 16.04 ubuntu. There is a suggestion I would like to make.

I installed netdata and did not change any config in the main file(/etc/netdata/netdata.conf).

Today, something misbehaved in my server(later I found it was some database query), and the server was not working properly. I was not able to log in to server(as the query was running and everything else stopped).

I checked in to see in my browser(netdata) that what app was misbehaving, but I saw that netdata stopped and it gave me 502.

Then I remembered that netdata's OOM value is pretty high and it must have been killed.

The point is : - 

Netdata is very light, consumes little cpu. I think it has very little effect overall when netdata is killed. But instead killing of netdata creates a problem. When server misbehaves, a person would like to see, which app is misbehaving, which One sees from netdata dashboard. So, if Netdata gets killed so early, how can the person see the status of server ?

It cost me almost 45 minutes due to this.

So, I suggest you should keep OOM low by default.

**Comment by Ferroin** [comment]
I would argue that netdata itself should not change it's OOM score, as this is really administrative policy and should be set by the init script or systemd unit file.  There are legitimate reasons to have netdata have a large OOM score though as it _does_ use a lot of _memory_ which is what the OOM killer is supposed to handle, and it's a non-critical piece of software for most systems.

**Comment by luvpreetsingh** [comment]
It may use a lot of memory. But the information its dashboard provides is beyond helpful and beyond good. So, one might not want to kill it when in trouble. It is the keyhole through which you can see everything.

**Comment by ktsaou** [comment]
netdata is under very active development.

One of my key concerns during this period, is to protect users from my faults. I spend quite some time tracking down possible memory leaks that could make netdata use unpredictable amounts of memory or high CPU. I literally hate such bugs and I consider them as totally unacceptable even during this very active development phase. netdata should always be stable after each and every PR.

But still, a high OOM score and a low CPU priority, by default, are a great line of defense for me. I try very hard on every single PR I am merging to eliminate such issues, but I am human, and there are times I work under pressure. The last thing I want is netdata to bring down your service due to a memory leak or an infinite loop I oversaw.

So, until we exit this very active development phase, I would prefer not to change these defaults.

Of course, I agree that eventually, netdata OOM score should by lower and CPU priority should be higher,  by default.

But please, respect my mental health :smile:. It is already quite a responsibility to have a complex piece of software like netdata, installed on every Linux/FreeBSD server out there and although I am not legally responsible for this, I consider it a very serious ethical responsibility.

**Comment by jpriebe** [comment]
Yesterday, we had an OOM episode, and one of our technicians restarted the process that was causing the problem, bringing the system back under control.  There was one more blip early this morning that I wanted to investigate, but netdata was down until I got to the office this morning.  :-(

@ktsaou - I think your cautious approach makes sense (even though my experience with netdata shows that it's a very stable, resource-efficient piece of software).  So I can wait until you feel comfortable changing the OOM score.

In the meantime, I'm thinking about setting up a cron job to try to restart netdata every X minutes in case it gets killed.  In a case where the situation resolves itself (or is resolved manually), we won't have to worry about netdata being down until somebody notices.

Or maybe there's a better workaround?

**Comment by ktsaou** [comment]
Thank you for your understanding!

There are a few far better ways. Pick any.

## /etc/netdata/netdata.conf

```
[global]
	OOM score = 1000
```

The system default is 0. The netdata default is 1000 (be the first to be killed). To disable killing netdata under any circumstance, use -1000.

It works like this:

1. The kernel calculates the `badness` of each process, ranging from 0 (never kill) to 1000 (always kill). For example, if a task is using all allowed memory, its badness score will be 1000.  If it is using half of its allowed memory, its score will be 500.

2. The value you set is added to the `badness` score before it is used to determine which task to kill.  Acceptable values range from -1000 to +1000.  This allows userspace to polarize the preference for oom killing either by always preferring a certain task or completely disabling it.  The lowest possible value, -1000, is equivalent to disabling oom killing entirely for that task since it will always report a badness score of 0.

## /etc/systemd/system/netdata.service

```
[Service]
Environment=OOMScoreAdjust=-1000
```

or replace the `ExecStart` line with this:

```
[Service]
ExecStart=/usr/sbin/netdata -D -W set global "OOM score" "-50"
```

**Comment by ktsaou** [comment]
You can check netdata `error.log` for the value it has set.

```sh
# grep -i "Out-Of-Memory" /var/log/netdata/error.log
2017-08-19 03:06:10: netdata INFO : Adjusted my Out-Of-Memory score to '1000'.
```

You can also check what the kernel has for netdata with this:

```sh
# cat "/proc/$(($(ps -e  | grep netdata | head -n 1 | cut -b 1-6)))/oom_score_adj"
1000
```

**Comment by jpriebe** [comment]
Thanks!  In 20 years of linux sysadmin work, I've never had to deal with Out-of-memory issues.  But now that we're using these tiny cloud-based machines, it seems to be something I have to worry about more...

**Comment by ktsaou** [comment]
Keep in mind the best way these days is to use cgroups to limit the resources per systemd service. You can limit CPU, disk IOPS, memory size, swap size, number of threads, etc **per service**.

Check this: https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html

**Comment by jpriebe** [comment]
I tried the netdata.conf technique to set the OOM score, and I don't think it's working:

```
$ grep -i "Out-Of-Memory" /opt/netdata/var/log/netdata/error.log
2017-08-22 12:25:36: netdata INFO : Adjusted my Out-Of-Memory score to '-1000'.

$ cat "/proc/$(($(ps -e  | grep netdata | head -n 1 | cut -b 1-6)))/oom_score_adj"
1000
```

Netdata 1.6.0, CentOS 7.3.1611 under AWS.  Installed via netdata-v1.6.0-549-g6dd6a777-x86_64-20170616-062819.gz.run

We're using a nightly because we needed the "send tags to opentsdb backend" feature, and we haven't updated to 1.7.0 yet.  If you think that 1.7.0 would solve this, I could try updating my systems.  Just takes some work to build new AMIs and deploy them.

**Comment by jpriebe** [comment]
I tried a few values for the OOM score in netdata.conf, and I found that negative numbers did not work.  

Value in netdata.conf: -1000
Resulting value in /proc/PID/oom_score_adj: 1000

Value in netdata.conf: -500
Resulting value in /proc/PID/oom_score_adj: 1000

Value in netdata.conf: 0
Resulting value in /proc/PID/oom_score_adj: 1

Value in netdata.conf: 10
Resulting value in /proc/PID/oom_score_adj: 10

**Comment by ktsaou** [comment]
Yes, I see exactly the same behaviour.

This is totally strange. I even tried `CAP_SYS_RESOURCE` as proc.txt suggests, but no luck. netdata can only increase the value...

**Comment by ktsaou** [comment]
The only solution I found to make it below zero is to edit `/etc/systemd/system/netdata.service` and set:

```
[Service]
OOMScoreAdjust=-1000
ExecStart=/usr/sbin/netdata -D -W set global "OOM score" keep
```

With the above, netdata runs with -1000.

**Comment by jpriebe** [comment]
I even tried using OOMScoreAdjust directly, per this article:

https://chuckyz.wordpress.com/2016/12/28/centos-7-disabling-oomkiller-for-a-process/

and that doesn't seem to work.  (I also set OOM score to the same value in netdata.conf so that netdata wouldn't try to set the oom score to a different value). 

I imagine that if I set the score to 1, that will be good enough to keep the OS from terminating netdata in most cases, since its memory footprint is small...

**Comment by ktsaou** [comment]
Have you tried changing `netdata.service` as I posted above? This file is not overwritten by netdata updates, so your changes will be permanent.

**Comment by jpriebe** [comment]
I was sending my comment before I saw yours.  I'm trying that now.

**Comment by jpriebe** [comment]
The netdata.service technique you found worked for me.  Thanks!

**Comment by ktsaou** [comment]
ok, I found the problem. netdata was not handling this case. The kernel is giving ENOPERM (permission denied) when an application is trying to lower its OOM score.
I have supplied default `netdata.service` and refactored the code to handle all the cases.
Also, netdata now sets its default OOM score to the one found running if it is not zero. So, by default, it adjusts its OOM score to 1000, only when it is zero. Of course if your change netdata.conf it will respect it.
